[LoopInterchange] Add simplifyLCSSAPhis: remove phi from non-exit bb

sjoerdmeijer opened the pull request:
This deals with a corner case of LCSSA phi nodes in the outer loop latch block: the loop was in LCSSA form, some transformations can come along (e.g. unswitch) and create an empty block:

     BB4:
       br label %BB5
     BB5:
       %old.cond.lcssa = phi i16 [ %cond, %BB4 ]
       br outer.header

Interchange then brings it in LCSSA form again and we get:

     BB4:
       %new.cond.lcssa = phi i16 [ %cond, %BB3 ]
       br label %BB5
     BB5:
       %old.cond.lcssa = phi i16 [ %new.cond.lcssa, %BB4 ]

Which means that we have a chain of LCSSA phi nodes from %new.cond.lcssa to %old.cond.lcssa. The problem is that interchange can reoder blocks BB4 and BB5 placing the use before the def if we don't check this. The solution is to simplify lcssa phis, and remove them from non-exit blocks if they are 1-input phi nodes. 

Fixes #160068

sjoerdmeijer committed:
[LoopInterchange] Also look at lcssa phis in outer loop latch block

This deals with a corner case of LCSSA phi nodes in the outer loop latch
block: the loop was in LCSSA form, some transformations can come along
(e.g. unswitch) and create an empty block:

     BB4:
       br label %BB5
     BB5:
       %old.cond.lcssa = phi i16 [ %cond, %BB4 ]
       br outer.header

Interchange then brings it in LCSSA form again and we get:

     BB4:
       %new.cond.lcssa = phi i16 [ %cond, %BB3 ]
       br label %BB5
     BB5:
       %old.cond.lcssa = phi i16 [ %new.cond.lcssa, %BB4 ]

Which means that we have a chain of LCSSA phi nodes from %new.cond.lcssa
to %old.cond.lcssa. The problem is that interchange can reoder blocks
BB4 and BB5 placing the use before the def if we don't check this. The
observation is that %old.cond.lcssa is unused, so instead of moving and
renaming these phi nodes, just delete it if it's trivially dead. If it
isn't trivially dead, it is handled elsewhere. The loop should still be
in LCSSA form, and if it isn't, formLCSSARecursively is called after the
interchange rewrite.

Fixes #160068

Meinersbur reviewed (commented):
If LoopInterchange has to build LCSSA form again anyway, have you considered cleaning this up at the same time? Then to be on the save side, do not any PHI nodes in `OuterLoopLatch` unless `InnerLoopExit == OuterLoopLatch` (like there is already a check `InnerLoopPreHeader != OuterLoopHeader` in `tightlyNested`). In addition to the case where `%old.cond.lcssa` is not trivally dead, I could imagine other dubious patterns, such as a PHI node referencing itself.

sjoerdmeijer committed:
Create new helper simplifyLCSSAPhis that checks single input phis in
outer loop latch blocks that are not exit blocks.

sjoerdmeijer commented:
I think I've addressed all comments. I've introduced a new helper that is now more intentional and explicit about checking non-exit blocks and single-input phis. I've added an assert there that I haven't managed to trigger in testing. I have changed the test case and made that a bit more interesting: the problematic phi was trivially dead before, but now it has a user in the exit block, which shows that we simplify the phi-chain and replace all uses.

Meinersbur commented on llvm/lib/Transforms/Scalar/LoopInterchange.cpp:
```diff
@@ -1872,6 +1873,50 @@ static void moveLCSSAPhis(BasicBlock *InnerExit, BasicBlock *InnerHeader,
   InnerLatch->replacePhiUsesWith(InnerLatch, OuterLatch);
 }
 
+// This deals with a corner case when a LCSSA phi node appears in a non-exit
```
```suggestion
/// This deals with a corner case when a LCSSA phi node appears in a non-exit
```

sjoerdmeijer replied:
The problem description is important, because it is not obvious and very subtle, so is worth explaining. I don't think one would get that from reading the implementation a small helper function. The implementation, this helper function, describes what this function performs, also with a few comments. But I don't disagree, so will add a sentence to these comments to make the solution explicit.

Meinersbur commented on llvm/test/Transforms/LoopInterchange/lcssa-phi-outer-latch.ll:
```diff
@@ -0,0 +1,75 @@
+BB2:
+  %j = phi i16 [ 0, %BB1 ], [ %j.next, %BB3 ]
+  %arrayidx.us.us = getelementptr i16, ptr null, i16 %j
+  %0 = load i16, ptr %arrayidx.us.us, align 1
```
UB in tests are unfortunately common, but LoopInterchange does not care. Could replace `null` in `getelementptr` with a funcation argument or a global

Meinersbur reviewed (commented):
@sjoerdmeijer 👍 This is what I meant. Can add assertion in `moveLCSSAPhis` ensuring that all those BBs are really empty.

@kasuga-fj We cannot just bail out in `moveLCSSAPhis`, when some changes have aleady been applied. If we check beforehand, we can also just remove the redundant PHI. 

LGTM, modulo the `null` dereference, future DA might not like that. Also consider (remaining) remarks from @kasuga-fj before landing.

kasuga-fj reviewed (commented):
> @kasuga-fj We cannot just bail out in `moveLCSSAPhis`, when some changes have aleady been applied. If we check beforehand, we can also just remove the redundant PHI.

I was imagining an implementation that adds an additional check (or replaces redundant PHIs) to `tightlyNested`, since you also mentioned a PHI which refers to itself. That said, the current approach also seems reasonable.

sjoerdmeijer commented on llvm/lib/Transforms/Scalar/LoopInterchange.cpp:
```diff
@@ -1882,6 +1927,9 @@ bool LoopInterchangeTransform::adjustLoopBranches() {
          InnerLoopPreHeader && "Guaranteed by loop-simplify form");
+
+  simplifyLCSSAPhis(OuterLoop, InnerLoop);
+
```
I was also wondering, and I have played with the place where to put this. But between this place here and `moveLCSSAPhis`, the CFG is massively rewired and sometimes in an funny state (i.e. it is under construction). I thought about doing this relatively early when the CFG is relatively stable before we start completely turning things around and adding and splitting things, and moving things around, which is here. Thus, I think it is fine here, and overall doesn't matter that much.

sjoerdmeijer committed:
Removed ndefined behaviour from the test, changed the comments slightly.

sjoerdmeijer reviewed (commented):
> I was imagining an implementation that adds an additional check (or replaces redundant PHIs) to tightlyNested, since you also mentioned a PHI which refers to itself. That said, the current approach also seems reasonable.

I think with the code snippet and check you suggested earlier, adding that to `thightlyNested`, this will result in a regression. I.e., I think it would result in the loop of the added test case being classified as not tightly nested. I thus think the current solution is fine, but we can revise this later should that be necessary. 

Thanks a lot @kasuga-fj and @Meinersbur for reviewing!
I saw a LGTM earlier. I have addressed the minor comments: removed the undefined behaviour from the test, changed the comments slightly, and replied to some queries. I will let this sit for a day, and merge this tomorrow if there are not further objections.

Meinersbur commented on llvm/test/Transforms/LoopInterchange/lcssa-phi-outer-latch.ll:
```diff
@@ -0,0 +1,75 @@
+  %j = phi i16 [ 0, %BB1 ], [ %j.next, %BB3 ]
+  %arrayidx.us.us = getelementptr i16, ptr null, i16 %j
+  %0 = load i16, ptr %arrayidx.us.us, align 1
+  %cond = select i1 false, i16 0, i16 0
```
> Yeah, interchange doesn't care,

Future DA might care, when checking AliasAnalysis on whether the base pointers themselves alias.

Meinersbur reviewed (approved):
Consider fixing using clang-format and adding an assert into `moveLCSSAPhis` whether all BBs from LcssaInnerExit to OuterLatch are really empty.

kasuga-fj reviewed (approved):
LGTM

kasuga-fj reviewed (commented):
I think it would be better to update the PR title.

sjoerdmeijer committed:
Clang-format
